### PR TITLE
Add json schema and cli lint command

### LINKED ISSUE #275: Add json-schema for ci files
It would be nice to add a json-schema for yml files to [schemastore.org](https://github.com/SchemaStore/schemastore). This would probably help many users with typos and autocompleting configs.

Before a schema can be added some kind of default path should be discussed see #133.

Maybe the following catalog entry would fit:
```json
{
      "name": "Woodpecker pipeline config",
      "description": "YAML schema for Woodpecker pipeline",
      "fileMatch": [
        ".woodpecker/**.yml",
        ".woodpecker.yml"
      ],
      "url": "https://json.schemastore.org/woodpecker.json"
    }
```

### PR DESCRIPTION
**Pull request by anbraten**
addresses #275 
preparation for #276 

# Changes
- add json schema file
- add tests to validate sample pipeline files
- add new command `lint` to cli to test a directory or single file to use correct schema

`woodpecker-cli lint ./pipeline/schema/.woodpecker/`
<details>
<summary>Sample output from `woodpecker-cli lint ./pipeline/schema/.woodpecker/`</summary>

```shell
# test-branches-exclude-include.yml
❌ Config is invalid
In branches: Must validate one and only one schema (oneOf)
In branches: Invalid type. Expected: array, given: object

# test-branches.yml
❌ Config is invalid
In branches: Must validate one and only one schema (oneOf)
In branches: Invalid type. Expected: array, given: object

# test-clone-skip.yml
❌ Config is invalid
In (root): Additional property skip_clone is not allowed

# test-clone.yml
✅ Config is valid

# test-matrix.yml
✅ Config is valid

# test-multi.yml
❌ Config is invalid
In (root): Additional property depends_on is not allowed

# test-plugin.yml
❌ Config is invalid
In (root): pipeline is required

# test-run-on.yml
❌ Config is invalid
In (root): Additional property run_on is not allowed

# test-service.yml
✅ Config is valid

# test-step.yml
❌ Config is invalid
In pipeline.detached: commands is required
In pipeline.volume.volumes: Invalid type. Expected: string, given: array

# test-when.yml
✅ Config is valid

# test-workspace.yml
✅ Config is valid
```
</details>

### CODE REVIEW

**Comment by 6543** [comment]
@anbraten please resolve conflict

**Comment by 6543** [comment]
```
./cli/release/linux/amd64/woodpecker-cli lint .woodpecker/
# docs.yml
✅ Config is valid

# helm.yml
✅ Config is valid

# main.yml
❌ Config is invalid
In pipeline.build-cli.commands: Invalid type. Expected: array, given: string
```

do we have a broken config?

**Comment by anbraten** [comment]
That's an error in the schema file. I didn't know / saw that a command is allowed to be just a single string instead of an array with a single item. I will adjust the schema.

File: pipeline/schema/schema.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	_ "embed"
+	"fmt"
+
+	"github.com/woodpecker-ci/woodpecker/shared/yml"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+//go:embed schema.json
```
Comment (6543) [inline pipeline/schema/schema.go:11]: Go  min version 1.16 - we should put a node somewhere

Comment (anbraten) [reply pipeline/schema/schema.go:11]: Maybe in the development guide (#314) with a list of required tools to develop woodpecker.

Comment (6543) [reply pipeline/schema/schema.go:11]: updated mentioned issue

**Comment by 6543** [comment]
one question are all schema in `pipeline/schema/.woodpecker/` valid by purpose?

**Comment by anbraten** [comment]
> one question are all schema in `pipeline/schema/.woodpecker/` valid by purpose?

Yes atm I only added valid configs for all possible options as the impossible options are infinitely large. Schemastore seems to have some [negative tests](https://github.com/SchemaStore/schemastore/tree/master/src) for their hosted schemas. I can try to add something similar.

**Comment by 6543** [comment]
yes just 1 or 2

**Review by 6543 — APPROVED** [review]

**Pull request merged**